jetbrains-core/tst/software/aws/toolkits/jetbrains/services/cwc/controller/ChatControllerTest.kt

```diff
@@ -0,0 +1,794 @@
+        }
+
+        // Assert
+        verify { mockJob.cancelChildren() }
+    }
+
+    @Test
+    fun `processChatItemVoted records telemetry`() {
+        val testMessage = IncomingCwcMessage.ChatItemVoted(
+            tabId = "testTabId",
+            messageId = "testMessageId",
+            vote = "testVote",
+        )
+
+        runBlocking {
+            chatController.processChatItemVoted(testMessage)
+        }
+
+        coVerify { telemetryHelper.recordInteractWithMessage(match { it == testMessage }) }
+    }
+
+    @Test
+    fun `processChatItemFeedback records telemetry`() {
+        val testMessage = IncomingCwcMessage.ChatItemFeedback(
+            tabId = "testTabId",
+            messageId = "testMessageId",
+            comment = "testComment",
+            selectedOption = "testSelectedOption",
+        )
+
+        runBlocking {
+            chatController.processChatItemFeedback(testMessage)
+        }
+
+        coVerify { telemetryHelper.recordInteractWithMessage(match { it == testMessage }) }
+    }
+
+    @Test
+    fun `processUIFocus records telemetry`() {
+        val testMessage = IncomingCwcMessage.UIFocus(
+            command = "testCommand",
+            type = FocusType.FOCUS
+        )
+
+        runBlocking {
+            chatController.processUIFocus(testMessage)
+        }
+
+        coVerify { telemetryHelper.recordEnterFocusChat() }
+    }
+
+    @Test
+    fun `processAuthFollowUpClick handles auth`() {
+        val testAuthType = AuthFollowUpType.FullAuth
+        val testMessage = IncomingCwcMessage.AuthFollowUpWasClicked(
+            tabId = "testTabId",
+            authType = testAuthType,
+        )
+
+        runBlocking {
+            chatController.processAuthFollowUpClick(testMessage)
+        }
+
+        coVerify { authController.handleAuth(any(), testAuthType) }
+    }
+
+    @Test
+    fun `processOnboardingPageInteraction handled`() {
+        // Arrange
+        val testTriggerId = "testTriggerId"
+
+        // mock context extractor
+        val mockContextExtractor = mockk<ActiveFileContextExtractor>(relaxed = true)
+        val testActiveFileContext = ActiveFileContext(null, null)
+        coEvery { mockContextExtractor.extractContextForTrigger(any()) } returns testActiveFileContext
+
+        // waitForTabId
+        val testTabId = "testTabId"
+        every { messagesFromUiToApp.flow } returns flowOf(
+            IncomingCwcMessage.TriggerTabIdReceived(testTriggerId, testTabId)
+        )
+
+        val testMessage = OnboardingPageInteraction(
+            OnboardingPageInteractionType.CwcButtonClick
+        )
+
+        runBlocking {
+            // Act
+            chatController.processOnboardingPageInteraction(testMessage, testTriggerId)
+        }
+
+        // Assert
+
+        // Check call to sendOnboardingPageInteractionMessage
+        coVerify {
+            messagePublisher.publish(
+                match {
```

Comment:
I would think about extracting some of these matchers into custom functions to make the tests more readable. Something like the following (not tested, just an outline):

```kotlin
fun MockKMatcherScope.matchOnboardingInteractionMessage(
  triggerId: String,
  message: StaticPrompt = StaticPrompt.OnboardingHelp,
  interactionType: OnboardingInteractionType = OnboardingInteractionType.CwcButtonClick
): AmazonQMessage = match {
  it is OnboardingPageInteractionMessage && ...
}

// usage
coVerify { messagePublisher.publish(matchOnboardingPageInteractionMessage(testTriggerId)) }
```

Reply:
Hmm unsure how this would reduce code or make things more readable?  

We have 8 coVerify tests on messagePublisher with around 4 different message types being matched on which each have different data schemas.  

Furthermore, we compare against actual test values in the code, if we were to consolidate it would still require creating a 'dummy' test message that we'd have to create and then match against in this helper method right?